src/connectors/assetpanda.py

```diff
@@ -0,0 +1,174 @@
+# on the user input data. We will reduce the fields down to just alpha numeric key strings so we can use
+# them as the keys in our final JSON data.
+def reduce_fields(accumulators: dict, field: dict) -> str:
+    # Take every word and join them by an underscore for create the new name
+    cleaner_name = "_".join(re.findall(r"[a-zA-Z]+", field["name"]))
+    field_key = field["key"]
+    accumulators[field_key] = cleaner_name
+    return accumulators
+
+
+# This method will iterate through the data and replace the data that matches against keys in the replace_keys
+# collection. It will convert data like the following example:
+# {"field_144": {"value": "00:0a:95:9d:68:16", "name": "MAC Address"}} => {"MAC_Address": "00:0a:95:9d:68:16"}
+def replace_device_key(list_device: list, replace_key: dict):
+    for key, value in replace_key.items():
+        for device in list_device:
+            if device.get(key, False):
+                if device.get(value) is not None:
+                    number = 2
+                    while device.get(f"{value}_{number}") is not None:
+                        number += 1
+                    value = f"{value}_{number}"
+                device[value] = device.pop(key)
+    return list_device
+
+
+### Main Functions ###
+
+def get_data(token: str, url: str, params: dict = {}) -> dict:
+    headers: dict = {
+        "Authorization": f"Bearer {token}"
+    }
+    try:
+        log.debug(f"Preparing GET: url={url} with params={params}")
+        req = requests.get(url, params=params, headers=headers)
+        req.raise_for_status()
+    except HTTPError as http_err:
+        log.error(f"Error GET: url={url}")
+        log.error(f"HTTP error occurred: {http_err}")
+        raise http_err
+    log.debug(req.status_code)
+
+    return req.json()
+    
+    
+def connect(connection_name, options):
+    landing_table = f'data.assetpanda_{connection_name}_connection ' # creates table in snowalert
+    
+    comment = yaml_dump(module='assetpanda', **options)
+
+    db.create_table(name=landing_table, cols=LANDING_TABLE_COLUMNS, comment=comment)
+
+    db.execute(f'GRANT INSERT, SELECT ON {landing_table} TO ROLE {SA_ROLE}')
+
+    return {
+        'newStage': 'finalized',
+        'newMessage': "AssetPanda ingestion tables created!",
+    }
+
+
+def ingest(table_name, options):
+    landing_table = f'data.{table_name}'
+    
+    token = options['token']
+    asset_entity_id = options['asset_entity_id']
+
+    general_url = f"https://api.assetpanda.com:443//v2/entities/{asset_entity_id}/objects"
+    fields_url  = f"https://api.assetpanda.com:443//v2/entities/{asset_entity_id}"
+
+    params = {
+        "offset": 0,
+        "limit": PAGE_SIZE,
+    }
+
+    total_object_count = 0 
+
+    insert_time = datetime.utcnow()
+
+    while params['offset'] <= total_object_count:
+
+        log.debug("total_object_count: ", total_object_count)
```

Comment:
Do we need to log this information every time we go through the loop? It's not changing, so this is going to produce a log line for every loop with the same information; this seems less than ideal.